File: simulator/src/node/roles/aggregator/simple_aggregator.cpp
#include <cstdint>
#include <memory>
#include <simgrid/s4u/Actor.hpp>
#include <simgrid/s4u/Engine.hpp>
#include <simgrid/s4u/Mailbox.hpp>
#include <variant>
#include <xbt/log.h>

#include "simple_aggregator.hpp"
#include "../../../protocol.hpp"
#include "../../../utils/utils.hpp"
#include "aggregator.hpp"


XBT_LOG_NEW_DEFAULT_CATEGORY(s4u_simple_aggregator, "Messages specific for this example");

using namespace std;
using namespace protocol;

SimpleAggregator::SimpleAggregator(std::unordered_map<std::string, std::string> *args, node_name name) : Aggregator(name)
{
    // Parsing arguments
    for (auto &[key, value]: *args)
    {
        switch (str2int(key.c_str()))
        {
            case str2int("is_main_aggregator"):
                bool ima = std::stoi(value);
                XBT_INFO("is_main_aggregator=%b", ima);
                this->is_main_aggregator = ima;
                break;
        }
    }

    delete args;
}

void SimpleAggregator::run()
{
    switch (this->state)
    {
        case INITIALIZING:
            {

                if (this->first_global_model)
                {
                    this->send_global_model();
                    this->first_global_model = false;
                }

                // Then wait for the event that tells us the number of connected client
                auto e = this->mc->get_nm_event();

                // If type of event is ClusterConnected it means that every node have been connected to us
                if (auto *conneted_event = get_if<Mediator::ClusterConnected>(e.get()))
                {
                    this->number_client_training = conneted_event->number_client_connected;
                    this->state = WAITING_LOCAL_MODELS;
                }
                break;
            }
        case WAITING_LOCAL_MODELS: 
            {
                auto op = this->mc->get_received_operation();

                // If the packet's operation is a SendLocalModel
                if (auto *op_send_local = get_if<operations::SendLocalModel>(op.get()))
                {
                    this->number_local_models += 1;
                    this->total_number_local_epochs += op_send_local->number_local_epochs_done;
                    XBT_INFO("nb local models: %lu", this->number_local_models);

                    if (this->number_local_models >= this->number_client_training)
                    {
                        this->state = AGGREGATING;
                    }
                }
                break;
            }
        case AGGREGATING:
            {
                this->aggregate();

                // Only check end condition as MainAggregator
                if (this->get_role_type() == NodeRole::MainAggregator 
                    && this->check_end_condition())
                {
                    this->print_end_report();
                    // Stop aggregating and send kills to the trainers
                    this->send_kills();
                }
                else
                {
                    this->send_global_model();
                    this->number_local_models = 0;
                    this->state = WAITING_LOCAL_MODELS;
                }
                break;
            }
    }
}

File: dmlogic/src/roles/aggregators/simple_aggregator.rs
use super::AggregatorBase;
use crate::moms::MOMEvent;
use crate::motherboard::{MotherboardEvent, MotherboardOrMOMEvent, MotherboardTask};
use crate::protocol::operations::Operation;
use crate::roles::{Role, RoleEvent};

// State enum for the SimpleAggregator, do not expose
enum State {
    Initializing,
    WaitingLocalModels,
    Aggregating,
}

pub struct SimpleAggregator {
    // Contains common fields and methods for all aggregators
    base: AggregatorBase,
    state: State,
    tasks: Vec<MotherboardTask>,
}

impl SimpleAggregator {
    pub fn new() -> SimpleAggregator {
        SimpleAggregator {
            base: AggregatorBase::new(),
            state: State::Initializing,
            tasks: Vec::new(),
        }
    }

    fn initializing(&mut self, event: MotherboardOrMOMEvent) -> Option<RoleEvent> {
        match event {
            // Case where we receive ClusterConnected event
            MotherboardOrMOMEvent::MOM(MOMEvent::ClusterConnected(nb_clients_connected)) => {
                self.base.number_client_training = nb_clients_connected;
                self.state = State::WaitingLocalModels;
                return Some(self.base.create_send_global_model_event());
            }
            _ => {} // Ignore all other events
        }

        None
    }

    fn waiting_local_models(&mut self, event: MotherboardOrMOMEvent) -> Option<RoleEvent> {
        match event {
            MotherboardOrMOMEvent::MOM(MOMEvent::OperationReceived(
                Operation::SendLocalModel {
                    number_local_epochs_done,
                },
            )) => {
                self.base.number_local_models += 1;
                self.base.total_number_local_epochs += number_local_epochs_done;

                if self.base.number_local_models >= self.base.number_client_training {
                    // Send the aggregation task and change state
                    self.tasks.push(self.base.create_aggregating_task());
                    self.state = State::Aggregating;
                }
            }
            _ => {} // Ignore all other events
        }

        None
    }

    fn aggregating(&mut self, event: MotherboardOrMOMEvent) -> Option<RoleEvent> {
        // TODO: Do we still need MainAggregator???? It's pretty annoying ngl
        // Only check end condition as MainAggregator
        // if (this->get_role_type() == NodeRole::MainAggregator
        //     && this->check_end_condition())
        // {
        //
        // Only check End condition or send global model when the aggregation task finished
        match event {
            MotherboardOrMOMEvent::Motherboard(MotherboardEvent::TaskExecDone) => {
                // Increment the number of aggregated models
                self.base.total_aggregated_models += self.base.number_local_models;
                // Compute the number of global epochs
                self.base.number_global_epochs =
                    self.base.total_aggregated_models / self.base.number_client_training;

                if self.base.check_end_condition() {
                    self.base.print_end_report();
                    // Return TBSP with Kill op
                    return Some(self.base.send_kills());
                } else {
                    self.base.number_local_models = 0;
                    self.state = State::WaitingLocalModels;
                    // Return TBSP with GlobalModel op
                    return Some(self.base.create_send_global_model_event());
                }
            }
            _ => {} // Ignore all other events
        }

        None
    }

    fn run_one_step(&mut self, event: MotherboardOrMOMEvent) -> Option<RoleEvent> {
        match self.state {
            State::Initializing => self.initializing(event),
            State::WaitingLocalModels => self.waiting_local_models(event),
            State::Aggregating => self.aggregating(event),
        }
    }
}

impl Role for SimpleAggregator {
    // fn get_role_type(&self) -> RoleEnum {
    //     self.base.get_role_type()
    // }

    fn put_event(&mut self, event: MotherboardOrMOMEvent) -> Option<RoleEvent> {
        self.run_one_step(event)
    }

    fn pop_task(&mut self) -> Option<MotherboardTask> {
        self.tasks.pop()
    }
}

#[cfg(test)]
mod tests {
    use crate::motherboard::KindExec;

    use super::*;

    #[test]
    fn test_typical_workflow() {
        let mut sa = SimpleAggregator::new();

        // It should be still in initializing mode
        assert!(matches!(sa.state, State::Initializing));

        // Pretend that 12 clients have connected
        let event = sa.put_event(MotherboardOrMOMEvent::MOM(MOMEvent::ClusterConnected(12)));

        // Should have 12 training clients and in WaitingLocalModels state
        assert!(sa.base.number_client_training == 12);
        assert!(matches!(sa.state, State::WaitingLocalModels));

        // The Aggregator sends a ToBeSentPacket event to the MOM
        assert!(matches!(event, Some(RoleEvent::ToBeSentPacket { .. })));

        for i in 0..12 {
            sa.put_event(MotherboardOrMOMEvent::MOM(MOMEvent::OperationReceived(
                Operation::SendLocalModel {
                    number_local_epochs_done: 3,
                },
            )));

            if i == 11 {
                // <- last local model
                // When all local models have been received, it should have changed its state to Aggregating
                assert!(matches!(sa.state, State::Aggregating));
                assert!(matches!(
                    sa.pop_task(),
                    Some(MotherboardTask::Exec(KindExec::Aggregating))
                ));
            } else {
                // Otherwise it should still be listening for new local models
                assert!(matches!(sa.state, State::WaitingLocalModels));
            }
        }

        sa.put_event(MotherboardOrMOMEvent::Motherboard(
            MotherboardEvent::TaskExecDone,
        ));
    }
}
